### PR TITLE
Test implementing loggging on one endpoint

### PR DESCRIPTION
**Pull request by notatiyyah**
## Link to Jira ticket

https://hackney.atlassian.net/browse/POH-327

### *What is the problem we're trying to solve*

We are having an error on the GET accounts by tenancy ID endpoint on the API. There is no logging on this API, which makes debugging very difficult.

### *What changes have we introduced*

- Update logging configuration
- Add basic logging to test that it comes through on AWS (more to come in follow up PRs)

#### _Checklist_

- [x] Added end-to-end (i.e. integration) tests where necessary e.g to test complete functionality of a newly added feature
- [x] Added tests to cover all new production code
- [x] Added comments to the README or updated relevant documentation _(add link to documentation)_, where necessary.
- [x] Checked all code for possible refactoring
- [x] Code pipeline builds correctly

### CODE REVIEW

File: AccountsApi/V1/Gateways/DynamoDbGateway.cs [hunk 1]
```diff
@@ -63,8 +67,10 @@ public async Task<List<Account>> GetAllArrearsAsync(AccountType accountType, str
             return data.Sort(sortBy, direction).ToList();
         }
 
+        [LogCall]
         public async Task<List<Account>> GetAllAsync(Guid targetId, AccountType accountType)
         {
+            _logger.LogDebug($"Calling _amazonDynamoDb.QueryAsync for targetId {targetId} and accountType {accountType}");
```
Comment (Duslerke) [inline AccountsApi/V1/Gateways/DynamoDbGateway.cs:73]: Same as with the GW comment: the `:` after the identifier names would help clear up confusion whilst the parameterised logged string values are missing for whatever reason.

File: AccountsApi/V1/Controllers/AccountApiController.cs [hunk 2]
```diff
@@ -89,6 +89,7 @@ public async Task<IActionResult> GetByIdAsync([FromRoute] Guid id)
         [ProducesResponseType(typeof(BaseErrorResponse), StatusCodes.Status400BadRequest)]
         [ProducesResponseType(typeof(BaseErrorResponse), StatusCodes.Status500InternalServerError)]
         [HttpGet]
+        [LogCall(LogLevel.Information)]
```
[the remark this answers is not in the document]
Comment (Duslerke) [reply AccountsApi/V1/Controllers/AccountApiController.cs:92]: Also we need this piece of work to cover all of the endpoints (EPs) within this API.
Despite us having experience "lack of logs" frustration with only one of EPs, it's best to get them all covered in one go - that way we'll have quicker response & debug time going forward should an issue arise anywhere.

[the remark this answers is not in the document]
Comment (notatiyyah) [reply AccountsApi/V1/Controllers/AccountApiController.cs:93]: Re: Logging the parameters
By default, we log the request url (which in this case includes the target id), so do you think it's necessary to have another log call with the same data?

image

File: AccountsApi/V1/Gateways/AccountSnsGateway.cs [hunk 3]
```diff
@@ -1,52 +0,0 @@
...
-            {
-                Message = message,
-                TopicArn = Environment.GetEnvironmentVariable("ACCOUNTS_SNS_ARN"),
-                MessageGroupId = "AccountSnsGroupId"
-            };
-            await _amazonSimpleNotificationService.PublishAsync(request).ConfigureAwait(false);
-        }
-    }
-}
-
```
Comment (Duslerke) [inline AccountsApi/V1/Gateways/AccountSnsGateway.cs:52]: Ehh... I don't think this is the PR to do this.

We're introducing all this risk of reconfiguring the SNS together with a harmless piece of work of adding more logs.

Maybe you could make this a separate PR instead? Let's first deploy the logging, and then do another PR for this.

Comment (Duslerke) [reply AccountsApi/V1/Gateways/AccountSnsGateway.cs:52]: Though it's good you spotted it & decided to change it.
However, it's best to not mix work A with work B for release history purposes, as well as risk management.

File: AccountsApi.Tests/LogCallAspectFixtureCollection.cs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+using Hackney.Core.Testing.Shared;
+using Xunit;
+
+namespace AccountsApi.Tests
+{
+    [CollectionDefinition("LogCall collection")]
+    public class LogCallAspectFixtureCollection : ICollectionFixture<LogCallAspectFixture>
+    { }
```
[the remark this answers is not in the document]
Comment (notatiyyah) [reply AccountsApi.Tests/LogCallAspectFixtureCollection.cs:8]: Here's the lbh-core documentation:
https://github.com/LBHackney-IT/lbh-core/blob/release/Hackney.Core/Hackney.Core.Testing/Hackney.Core.Testing.Shared/LogCallAspectFixture.cs

Essentially it configures a mock logger to use in the tests, and they fail without it.

File: AccountsApi/V1/Controllers/AccountApiController.cs [hunk 5]
```diff
@@ -219,6 +223,7 @@ public async Task<IActionResult> Patch([FromRoute] Guid id, [FromBody] JsonPatch
         [ProducesResponseType(typeof(BaseErrorResponse), StatusCodes.Status500InternalServerError)]
         [Route("arrears")]
         [HttpGet]
+        [LogCall(LogLevel.Information)]
```
[the remark this answers is not in the document]
Comment (notatiyyah) [reply AccountsApi/V1/Controllers/AccountApiController.cs:226]: Leaving it as `[Logcall]` defaults to`LogLevel.Trace`, which is good for catching exceptions etc, but we don't want the controllers to return the entire stack trace because it would be a security issue(?).

I believe the log level is changed to `Information` in the controller so that if it returns an error, it's just a short error message, not the whole stack trace, but we'll still be able to see the whole stack trace in Cloudwatch

File: AccountsApi/AccountsApi.csproj [hunk 6]
```diff
@@ -29,6 +29,7 @@
     <PackageReference Include="AWSXRayRecorder.Handlers.EntityFramework" Version="1.1.0" />
     <PackageReference Include="Hackney.Core.ElasticSearch" Version="1.49.0" />
     <PackageReference Include="Hackney.Core.Logging" Version="1.30.0" />
+    <PackageReference Include="hackney.core.sns" Version="1.52.0" />
```
[the remark this answers is not in the document]
Comment (notatiyyah) [reply AccountsApi/AccountsApi.csproj]: Yep, didn't need it, it's been removed

File: AccountsApi.Tests/LogCallAspectFixtureCollection.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Duslerke) [reply AccountsApi.Tests/LogCallAspectFixtureCollection.cs:8]: Thanks!

File: AccountsApi/V1/Controllers/AccountApiController.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Duslerke) [reply AccountsApi/V1/Controllers/AccountApiController.cs:226]: Would be good to try and see what logs we'd get if the exception was thrown within the factory, or within a use case of the gateway. The question is: "if the error happens anywhere but within the controller method directly, do we lose information of where it happened?".

It shouldn't be a security issue as the logs get written to CW. So only the devs with CW permissions will be able to view them. But even then, I don't think there's anything sensitive within that stack trace. I suppose the stack trace reveals the structure of your code somewhat, but... the API's source code itself is public.

The stack trace could be useful - would help to determine whether the error happened:
1. whilst model binding,
2. within the controller method
3. within the helper method that controller called
4. within the injected dependency method that controller called.

[the remark this answers is not in the document]
Comment (Duslerke) [reply AccountsApi/V1/Controllers/AccountApiController.cs:226]: It's probably fine and no change is likely needed, but would be good to deploy a version of the app, _where it throws an exception within some method that controller calls_, to see what we get back within logs, so we could be confident that we're getting enough to debug crashes.

[the remark this answers is not in the document]
Comment (Duslerke) [reply AccountsApi/V1/Controllers/AccountApiController.cs:226]: That's exactly what we need - so we're good to go with it as is then.

File: AccountsApi.Tests/V1/Gateways/DynamoDbGatewayTests.cs [hunk 7]
```diff
@@ -44,10 +45,11 @@ public async Task GetById_DbReturnsNull_ReturnsNull()
         {
             _dynamoDb.Setup(_ => _.LoadAsync<AccountDbEntity>(It.IsAny<Guid>(), It.IsAny<CancellationToken>()))
                 .ReturnsAsync((AccountDbEntity) null);
-
-            var response = await _gateway.GetByIdAsync(Guid.NewGuid()).ConfigureAwait(false);
+            var id = Guid.NewGuid();
+            var response = await _classUnderTest.GetByIdAsync(id).ConfigureAwait(false);
 
             response.Should().BeNull();
+            _logger.VerifyExact(LogLevel.Debug, $"Calling _dynamoDbContext.LoadAsync for id: {id}", Times.Once());
```
Comment (Duslerke) [inline AccountsApi.Tests/V1/Gateways/DynamoDbGatewayTests.cs:52]: We probably don't need this anymore given there're dedicated tests now.

File: AccountsApi.Tests/V1/Gateways/DynamoDbGatewayTests.cs [hunk 8]
```diff
@@ -154,13 +173,15 @@ public async Task Update_WithValidModel_WorksOnce()
...
+            await _classUnderTest.UpdateAsync(domain).ConfigureAwait(false);
 
             _dynamoDb.Verify(_ => _.SaveAsync(It.IsAny<AccountDbEntity>(), default), Times.Once);
-            var updatedDomain = await _gateway.GetByIdAsync(domain.Id).ConfigureAwait(false);
+            var updatedDomain = await _classUnderTest.GetByIdAsync(domain.Id).ConfigureAwait(false);
             updatedDomain.Should().NotBeNull();
             updatedDomain.Should().BeEquivalentTo(domain,
                 opt => opt.Excluding(f => f.LastUpdatedAt));
+
+            _logger.VerifyExact(LogLevel.Debug, $"Calling _dynamoDbContext.SaveAsync for account: {domain}", Times.Once());
```
[the remark this answers is not in the document]
Comment (Duslerke) [reply AccountsApi.Tests/V1/Gateways/DynamoDbGatewayTests.cs:254]: We don't need this logger mock assertion given you've added dedicated loging tests.

File: AccountsApi.Tests/V1/Gateways/DynamoDbGatewayTests.cs [hunk 9]
```diff
@@ -184,10 +205,11 @@ public async Task GetAllArrears_DbReturnsZeroItems_ReturnsEmptyList()
             _amazonDynamoDb.Setup(_ => _.QueryAsync(It.IsAny<QueryRequest>(), It.IsAny<CancellationToken>()))
                 .ReturnsAsync(new QueryResponse());
 
-            var result = await _gateway.GetAllArrearsAsync(AccountType.Master, "Field", Direction.Asc).ConfigureAwait(false);
+            var result = await _classUnderTest.GetAllArrearsAsync(AccountType.Master, "Field", Direction.Asc).ConfigureAwait(false);
 
             Assert.NotNull(result);
             Assert.Empty(result);
+            _logger.VerifyExact(LogLevel.Debug, $"Calling _amazonDynamoDb.QueryAsync for accountType: {AccountType.Master}", Times.Once());
```
[the remark this answers is not in the document]
Comment (Duslerke) [reply AccountsApi.Tests/V1/Gateways/DynamoDbGatewayTests.cs:309]: You can remove this logger assertion ... as you've added the dedicated tests bellow

File: AccountsApi.Tests/V1/Gateways/DynamoDbGatewayTests.cs [hunk 10]
```diff
@@ -199,11 +221,12 @@ public async Task GetAllArrears_DbReturnsItems_ReturnsSortedList()
             _amazonDynamoDb.Setup(_ => _.QueryAsync(It.IsAny<QueryRequest>(), It.IsAny<CancellationToken>()))
                 .ReturnsAsync(response);
 
-            var result = await _gateway.GetAllArrearsAsync(AccountType.Master, "Field", Direction.Asc).ConfigureAwait(false);
+            var result = await _classUnderTest.GetAllArrearsAsync(AccountType.Master, "Field", Direction.Asc).ConfigureAwait(false);
 
             result.Should().NotBeNull();
             result.Should().BeEquivalentTo(response.ToAccounts());
             Assert.All(result, item => item.Should().NotBeNull());
+            _logger.VerifyExact(LogLevel.Debug, $"Calling _amazonDynamoDb.QueryAsync for accountType: {AccountType.Master}", Times.Once());
```
[the remark this answers is not in the document]
Comment (Duslerke) [reply AccountsApi.Tests/V1/Gateways/DynamoDbGatewayTests.cs:326]: This logging assertion can be removed now as you the exact same one within the dedicated tests bellow.

**Review by Duslerke — COMMENTED** [review]
Good work! Nearly there.

All that's left is to clean up the redundant Logging Assertions for which you've created dedicated tests.

File: AccountsApi.Tests/V1/Gateways/DynamoDbGatewayTests.cs [hunk 11]
```diff
@@ -108,9 +112,14 @@ public async Task GetAll_DbReturnsZeroItems_ReturnsEmptyList()
...
 
-            var result = await _gateway.GetAllAsync(Guid.NewGuid(), AccountType.Master).ConfigureAwait(false);
+            var targetId = Guid.NewGuid();
+            var accountType = AccountType.Master;
+
+            var result = await _classUnderTest.GetAllAsync(targetId, accountType).ConfigureAwait(false);
             result.Should().NotBeNull();
             result.Should().HaveCount(0);
+
+            _logger.VerifyExact(LogLevel.Debug, $"Calling _amazonDynamoDb.QueryAsync for targetId: {targetId} and accountType: {accountType}", Times.Once());
```
[the remark this answers is not in the document]
Comment (Duslerke) [reply AccountsApi.Tests/V1/Gateways/DynamoDbGatewayTests.cs]: The `_classUnderTest.GetAllAsync(...)`  does not have dedicated logging tests, so without adding those in, this assertion cannot be removed.

File: AccountsApi.Tests/V1/Gateways/DynamoDbGatewayTests.cs [hunk 12]
```diff
@@ -121,10 +130,15 @@ public async Task GetAll_DbReturnsItems_ReturnsList()
...
-            var result = await _gateway.GetAllAsync(Guid.NewGuid(), AccountType.Master).ConfigureAwait(false);
+            var targetId = Guid.NewGuid();
+            var accountType = AccountType.Master;
+
+            var result = await _classUnderTest.GetAllAsync(targetId, accountType).ConfigureAwait(false);
             _amazonDynamoDb.Verify(x => x.QueryAsync(It.IsAny<QueryRequest>(), It.IsAny<CancellationToken>()), Times.Once);
             result.Should().NotBeNull();
             result.Should().BeEquivalentTo(response.ToAccounts());
+
+            _logger.VerifyExact(LogLevel.Debug, $"Calling _amazonDynamoDb.QueryAsync for targetId: {targetId} and accountType: {accountType}", Times.Once());
```
[the remark this answers is not in the document]
Comment (Duslerke) [reply AccountsApi.Tests/V1/Gateways/DynamoDbGatewayTests.cs]: Same as above - the `_classUnderTest.GetAllAsync(...)`  does not have dedicated logging tests, so without adding those in, this assertion cannot be removed.

**Review by Duslerke — APPROVED** [review]
